test/Effect.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE DeriveFunctor, ExistentialQuantification, FlexibleInstances, GeneralizedNewtypeDeriving, MultiParamTypeClasses, StandaloneDeriving, TypeFamilies, TypeOperators, UndecidableInstances #-}
+module Effect
+( tests
+) where
+
+import Control.Algebra
+import Control.Carrier.NonDet.Church
+import Control.Carrier.State.Strict
+import Control.Monad.Trans.Class
+import Data.Functor.Identity
+import Data.Monoid (Sum(..))
+import Prelude hiding (all)
+import Test.Tasty
+import Test.Tasty.HUnit
+
+tests :: TestTree
+tests = testGroup "Effect"
+  [ testCase "allows effects to constrain the state functor" $ do
+    let bs = run . runAll . runState (sum 0) $ do
+          s <- pure "hello" <|> pure "world"
+          bs <- all (foldMapA (\ (c, i) -> c <$ put (sum i)) (zip s [1..]))
+          bs <$ put (sum 1) <|> pure bs
+    bs @?= [(Sum 1, "hello"), (Sum 15, "hello"), (Sum 1, "world"), (Sum 15, "world")]
+  ] where
+  sum i = Sum (i :: Int)
```

Comment:
There’s probably some reasonable property to be distilled out of this, but I didn’t feel it was worth the effort given that this `All` effect isn’t a feature of the library.